dask/dataframe/io/tests/test_orc.py

```diff
@@ -75,8 +78,104 @@ def test_orc_single(orc_files):
 
 
 def test_orc_multiple(orc_files):
-    d = read_orc(orc_files[0])
-    d2 = read_orc(orc_files)
+    d = dd.read_orc(orc_files[0])
+    d2 = dd.read_orc(orc_files)
     assert_eq(d2[columns], dd.concat([d, d])[columns], check_index=False)
-    d2 = read_orc(os.path.dirname(orc_files[0]) + "/*.orc")
+    d2 = dd.read_orc(os.path.dirname(orc_files[0]) + "/*.orc")
     assert_eq(d2[columns], dd.concat([d, d])[columns], check_index=False)
+
+
+@pytest.mark.skipif(
+    parse_version(pa.__version__) < parse_version("4.0.0"),
+    reason=("PyArrow>=4.0.0 required for ORC write support."),
+)
+@pytest.mark.parametrize("index", [None, "i32"])
+@pytest.mark.parametrize("columns", [None, ["i32", "i64", "f"]])
+def test_orc_roundtrip(tmpdir, index, columns):
+    tmp = str(tmpdir)
+    data = pd.DataFrame(
+        {
+            "i32": np.arange(1000, dtype=np.int32),
+            "i64": np.arange(1000, dtype=np.int64),
+            "f": np.arange(1000, dtype=np.float64),
+            "bhello": np.random.choice(["hello", "yo", "people"], size=1000).astype(
+                "O"
+            ),
+        }
+    )
+    if index:
+        data.set_index(index, inplace=True)
+    df = dd.from_pandas(data, chunksize=500)
+    if columns:
+        data = data[[c for c in columns if c != index]]
+
+    # Write
+    df.to_orc(tmp, write_index=bool(index))
+
+    # Read
+    df2 = dd.read_orc(tmp, index=index, columns=columns)
+    assert_eq(data, df2, check_index=bool(index))
+
+
+@pytest.mark.skipif(
+    parse_version(pa.__version__) < parse_version("4.0.0"),
+    reason=("PyArrow>=4.0.0 required for ORC write support."),
+)
+@pytest.mark.parametrize("split_stripes", [True, False, 2, 4])
+def test_orc_roundtrip_aggregate_files(tmpdir, split_stripes):
+    tmp = str(tmpdir)
+    data = pd.DataFrame(
+        {
+            "a": np.arange(100, dtype=np.float64),
+            "b": np.random.choice(["cat", "dog", "mouse"], size=100),
+        }
+    )
+    df = dd.from_pandas(data, npartitions=8)
+    df.to_orc(tmp, write_index=False)
+    df2 = dd.read_orc(tmp, split_stripes=split_stripes, aggregate_files=True)
+
+    # Check that we have the expected partition count
+    # and that the data is correct
+    if split_stripes:
+        assert df2.npartitions == df.npartitions / int(split_stripes)
+    else:
+        assert df2.npartitions == df.npartitions
+    assert_eq(data, df2, check_index=False)
+
+
+def test_orc_roundtrip_aggregate_files_offset(tmpdir):
+
+    # PyORC required to write orc files with a specific
+    # stripe sizes
+    pyorc = pytest.importorskip("pyorc")
```

Comment:
I don't have plans to maintain a pyorc engine myself, but it seems that we require a library like this to write ORC files with custom stripe sizes.

Reply:
@jrbourbeau - I'm interested to know if you have thoughts on using pyorc for testing purposes only here.  

**Some Notes**:

We are testing the case where the ORC dataset is comprised of multiple 2-stripe files, and the user sets something like `split_stripes=3` and `aggregate_files=True`. In this case, we  will aggregate the first file together with the first stripes of the second file to generate the first Dask-DataFrame partition.

It is impossible to test this "offset file-aggregation" use case unless we are able to write ORC files comprising multiple stripes.  Since pyarrow does not provide a public python API for setting the stripe size, we would need to write **very large** files to use pyarrow for this test.

Reply:
> I'm interested to know if you have thoughts on using pyorc for testing purposes only here

No strong opinion from me. `pyorc` is a relatively small package that's available on PyPI and conda-forge. IIUC it sounds like the ability to set the stripe size is in scope for pyarrow, but just not implemented yet (cc @jorisvandenbossche). If that's the case, we might want to add a comment to this test which mentions that we want to move to using pyarrow once this feature is available. 

Additionally we should add `pyorc` to our CI environment files so this test is run

Reply:
Cool - I added a comment to the test and also added pyorc to CI for the cases where pyarrow-4+ may be picked up (otherwise the pyorc test would be skipped anyway)

Reply:
Thanks Joris! Added a JIRA issue [here](https://issues.apache.org/jira/browse/ARROW-13571)